controller/program/src/instructions/start_rebalance.rs

```diff
@@ -0,0 +1,491 @@
+    let inp_lst_state = unsafe { inp_lst_state.as_lst_state() };
+
+    if inp_lst_state.is_input_disabled != 0 {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::LstInputDisabled).into());
+    }
+
+    verify_end_rebalance_exists(
+        instructions_acc.data(),
+        abr.get(*ix_prefix.inp_lst_mint()).key(),
+    )?;
+
+    let out_lst_mint_acc = abr.get(*ix_prefix.out_lst_mint());
+    let out_token_prog = out_lst_mint_acc.owner();
+
+    let inp_lst_mint_acc = abr.get(*ix_prefix.inp_lst_mint());
+    let inp_token_prog = inp_lst_mint_acc.owner();
+
+    let expected_out_reserves = create_raw_pool_reserves_addr(
+        out_token_prog,
+        &out_lst_state.mint,
+        &out_lst_state.pool_reserves_bump,
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    let expected_inp_reserves = create_raw_pool_reserves_addr(
+        inp_token_prog,
+        &inp_lst_state.mint,
+        &inp_lst_state.pool_reserves_bump,
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    let expected_pks = NewStartRebalanceIxPreAccsBuilder::start()
+        .with_rebalance_auth(&pool.rebalance_authority)
+        .with_pool_state(&POOL_STATE_ID)
+        .with_lst_state_list(&LST_STATE_LIST_ID)
+        .with_rebalance_record(&REBALANCE_RECORD_ID)
+        .with_out_lst_mint(&out_lst_state.mint)
+        .with_inp_lst_mint(&inp_lst_state.mint)
+        .with_out_pool_reserves(&expected_out_reserves)
+        .with_inp_pool_reserves(&expected_inp_reserves)
+        .with_withdraw_to(abr.get(*ix_prefix.withdraw_to()).key())
+        .with_instructions(&INSTRUCTIONS_SYSVAR_ID)
+        .with_system_program(&SYSTEM_PROGRAM_ID)
+        .with_out_lst_token_program(out_token_prog)
+        .build();
+    verify_pks(abr, &ix_prefix.0, &expected_pks.0)?;
+
+    verify_signers(abr, &ix_prefix.0, &START_REBALANCE_IX_PRE_IS_SIGNER.0)
+        .map_err(|expected_signer| log_and_return_acc_privilege_err(abr, *expected_signer))?;
+
+    verify_not_rebalancing_and_not_disabled(pool)?;
+
+    let out_calc_accs_len = args.out_lst_value_calc_accs as usize;
+    if out_calc_accs_len == 0 {
+        return Err(NOT_ENOUGH_ACCOUNT_KEYS.into());
+    }
+    if suf.len() < out_calc_accs_len + 1 {
+        return Err(NOT_ENOUGH_ACCOUNT_KEYS.into());
+    }
+
+    let (out_calc_prog, out_suf) = suf.split_first().ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+    let (out_calc, inp_suf) = out_suf.split_at(out_calc_accs_len - 1);
+
+    let (inp_calc_prog, inp_calc) = inp_suf.split_first().ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    verify_sol_value_calculator_is_program(abr.get(*out_calc_prog))?;
+    verify_sol_value_calculator_is_program(abr.get(*inp_calc_prog))?;
+
+    verify_pks(
+        abr,
+        &[*out_calc_prog],
+        &[&out_lst_state.sol_value_calculator],
+    )?;
+    verify_pks(
+        abr,
+        &[*inp_calc_prog],
+        &[&inp_lst_state.sol_value_calculator],
+    )?;
```

Comment:
id like you to read this thread as well: https://github.com/igneous-labs/inf-1.5/pull/77#discussion_r2450730947